python/cuml/ensemble/randomforest_shared.pyx

```diff
@@ -0,0 +1,132 @@
+#
+# Copyright (c) 2019-2020, NVIDIA CORPORATION.
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#     http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+#
+
+# cython: profile=False
+# distutils: language = c++
+# cython: embedsignature = True
+# cython: language_level = 3
+
+from libcpp.vector cimport vector
+from cython.operator cimport dereference as deref, preincrement as inc
+from cpython.object cimport PyObject
+from libc.stdint cimport uintptr_t
+from typing import Tuple, Dict, List, Union
+
+cdef extern from "treelite/tree.h" namespace "treelite":
+    cdef struct PyBufferFrame:
+        void* buf
+        char* format
+        size_t itemsize
+        size_t nitem
+    cdef cppclass Model:
+        vector[PyBufferFrame] GetPyBuffer() except +
+        void InitFromPyBuffer(vector[PyBufferFrame] frames) except +
+
+cdef extern from "Python.h":
+    Py_buffer* PyMemoryView_GET_BUFFER(PyObject* mview)
+
+cdef class PyBufferFrameWrapper:
+    cdef PyBufferFrame _handle
+    cdef Py_ssize_t shape[1]
+    cdef Py_ssize_t strides[1]
+
+    def __cinit__(self):
+        pass
+
+    def __dealloc__(self):
+        pass
+
+    def __getbuffer__(self, Py_buffer* buffer, int flags):
+        cdef Py_ssize_t itemsize = self._handle.itemsize
+
+        self.shape[0] = self._handle.nitem
+        self.strides[0] = itemsize
+
+        buffer.buf = self._handle.buf
+        buffer.format = self._handle.format
+        buffer.internal = NULL
+        buffer.itemsize = itemsize
+        buffer.len = self._handle.nitem * itemsize
+        buffer.ndim = 1
+        buffer.obj = self
+        buffer.readonly = 0
+        buffer.shape = self.shape
+        buffer.strides = self.strides
+        buffer.suboffsets = NULL
+
+    def __releasebuffer__(self, Py_buffer *buffer):
+        pass
+
+cdef PyBufferFrameWrapper MakePyBufferFrameWrapper(PyBufferFrame handle):
+    cdef PyBufferFrameWrapper wrapper = PyBufferFrameWrapper()
+    wrapper._handle = handle
+    return wrapper
+
+cdef list _get_frames(ModelHandle model):
+    frames = []
+    cdef Model* model_obj = <Model*>model
+    cdef vector[PyBufferFrame] interface = model_obj.GetPyBuffer()
+    cdef vector[PyBufferFrame].iterator it = interface.begin()
+    while it != interface.end():
+        v = deref(it)
+        w = MakePyBufferFrameWrapper(v)
+        frames.append(memoryview(w))
+        inc(it)
+    return frames
+
+cdef ModelHandle _init_from_frames(vector[PyBufferFrame] frames) except *:
+    cdef Model* model_obj = new Model()
```

Comment:
How do these get cleaned up?

Reply:
I'm concerned about memory leak as well. Looking at #2275

Reply:
FWIW in RMM and cuDF we have managed this by using `std::unique_ptr` to hold objects like this in Cython Extension Types. Here's an example from [RMM's `DeviceBuffer`]( https://github.com/rapidsai/rmm/blob/eb446bb37313a26c6cd0947b38b63496ac382d5a/python/rmm/_lib/device_buffer.pxd#L41-L42 ).

Reply:
I merged latest branch-0.14 which contains a fix for the memory leak: #2282